### ISSUE TITLE
h5netcdf creates files with groups that cannot be read with netCDF4

### ISSUE THREAD

**Issue by shoyer** [shoyer/h5netcdf#55]
As reported over in https://github.com/pydata/xarray/issues/2974.

For reference, the output of h5dump for files created with netcdf4 and h5netcdf:
```
HDF5 "test-netcdf4.nc" {
GROUP "/" {
   ATTRIBUTE "_NCProperties" {
      DATATYPE  H5T_STRING {
         STRSIZE 34;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_ASCII;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SCALAR
      DATA {
      (0): "version=2,netcdf=4.6.2,hdf5=1.10.4"
      }
   }
   GROUP "grp1" {
      DATASET "data1" {
         DATATYPE  H5T_STD_I64LE
         DATASPACE  SIMPLE { ( 2, 2 ) / ( 2, 2 ) }
         DATA {
         (0,0): 1, 2,
         (1,0): 3, 4
         }
         ATTRIBUTE "DIMENSION_LIST" {
            DATATYPE  H5T_VLEN { H5T_REFERENCE { H5T_STD_REF_OBJECT }}
            DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
            DATA {
            (0): (DATASET 566 /grp1/y ), (DATASET 876 /grp1/x )
            }
         }
         ATTRIBUTE "_Netcdf4Dimid" {
            DATATYPE  H5T_STD_I32LE
            DATASPACE  SCALAR
            DATA {
            (0): 0
            }
         }
      }
      DATASET "x" {
         DATATYPE  H5T_STD_I64LE
         DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
         DATA {
         (0): 1, 2
         }
         ATTRIBUTE "CLASS" {
            DATATYPE  H5T_STRING {
               STRSIZE 16;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_ASCII;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SCALAR
            DATA {
            (0): "DIMENSION_SCALE"
            }
         }
         ATTRIBUTE "NAME" {
            DATATYPE  H5T_STRING {
               STRSIZE 2;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_ASCII;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SCALAR
            DATA {
            (0): "x"
            }
         }
         ATTRIBUTE "REFERENCE_LIST" {
            DATATYPE  H5T_COMPOUND {
               H5T_REFERENCE { H5T_STD_REF_OBJECT } "dataset";
               H5T_STD_I32LE "dimension";
            }
            DATASPACE  SIMPLE { ( 1 ) / ( 1 ) }
            DATA {
            (0): {
                  DATASET 3294 /grp1/data1 ,
                  1
               }
            }
         }
         ATTRIBUTE "_Netcdf4Dimid" {
            DATATYPE  H5T_STD_I32LE
            DATASPACE  SCALAR
            DATA {
            (0): 1
            }
         }
      }
      DATASET "y" {
         DATATYPE  H5T_STD_I64LE
         DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
         DATA {
         (0): 1, 2
         }
         ATTRIBUTE "CLASS" {
            DATATYPE  H5T_STRING {
               STRSIZE 16;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_ASCII;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SCALAR
            DATA {
            (0): "DIMENSION_SCALE"
            }
         }
         ATTRIBUTE "NAME" {
            DATATYPE  H5T_STRING {
               STRSIZE 2;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_ASCII;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SCALAR
            DATA {
            (0): "y"
            }
         }
         ATTRIBUTE "REFERENCE_LIST" {
            DATATYPE  H5T_COMPOUND {
               H5T_REFERENCE { H5T_STD_REF_OBJECT } "dataset";
               H5T_STD_I32LE "dimension";
            }
            DATASPACE  SIMPLE { ( 1 ) / ( 1 ) }
            DATA {
            (0): {
                  DATASET 3294 /grp1/data1 ,
                  0
               }
            }
         }
         ATTRIBUTE "_Netcdf4Dimid" {
            DATATYPE  H5T_STD_I32LE
            DATASPACE  SCALAR
            DATA {
            (0): 0
            }
         }
      }
   }
   GROUP "grp2" {
      DATASET "data2" {
         DATATYPE  H5T_STD_I64LE
         DATASPACE  SIMPLE { ( 3, 3 ) / ( 3, 3 ) }
         DATA {
         (0,0): 1, 2, 3,
         (1,0): 4, 5, 6,
         (2,0): 7, 8, 9
         }
         ATTRIBUTE "DIMENSION_LIST" {
            DATATYPE  H5T_VLEN { H5T_REFERENCE { H5T_STD_REF_OBJECT }}
            DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
            DATA {
            (0): (DATASET 9761 /grp2/y ), (DATASET 10071 /grp2/x )
            }
         }
      }
      DATASET "x" {
         DATATYPE  H5T_STD_I64LE
         DATASPACE  SIMPLE { ( 3 ) / ( 3 ) }
         DATA {
         (0): 1, 2, 3
         }
         ATTRIBUTE "CLASS" {
            DATATYPE  H5T_STRING {
               STRSIZE 16;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_ASCII;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SCALAR
            DATA {
            (0): "DIMENSION_SCALE"
            }
         }
         ATTRIBUTE "NAME" {
            DATATYPE  H5T_STRING {
               STRSIZE 2;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_ASCII;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SCALAR
            DATA {
            (0): "x"
            }
         }
         ATTRIBUTE "REFERENCE_LIST" {
            DATATYPE  H5T_COMPOUND {
               H5T_REFERENCE { H5T_STD_REF_OBJECT } "dataset";
               H5T_STD_I32LE "dimension";
            }
            DATASPACE  SIMPLE { ( 1 ) / ( 1 ) }
            DATA {
            (0): {
                  DATASET 12497 /grp2/data2 ,
                  1
               }
            }
         }
         ATTRIBUTE "_Netcdf4Dimid" {
            DATATYPE  H5T_STD_I32LE
            DATASPACE  SCALAR
            DATA {
            (0): 3
            }
         }
      }
      DATASET "y" {
         DATATYPE  H5T_STD_I64LE
         DATASPACE  SIMPLE { ( 3 ) / ( 3 ) }
         DATA {
         (0): 1, 2, 3
         }
         ATTRIBUTE "CLASS" {
            DATATYPE  H5T_STRING {
               STRSIZE 16;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_ASCII;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SCALAR
            DATA {
            (0): "DIMENSION_SCALE"
            }
         }
         ATTRIBUTE "NAME" {
            DATATYPE  H5T_STRING {
               STRSIZE 2;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_ASCII;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SCALAR
            DATA {
            (0): "y"
            }
         }
         ATTRIBUTE "REFERENCE_LIST" {
            DATATYPE  H5T_COMPOUND {
               H5T_REFERENCE { H5T_STD_REF_OBJECT } "dataset";
               H5T_STD_I32LE "dimension";
            }
            DATASPACE  SIMPLE { ( 1 ) / ( 1 ) }
            DATA {
            (0): {
                  DATASET 12497 /grp2/data2 ,
                  0
               }
            }
         }
         ATTRIBUTE "_Netcdf4Dimid" {
            DATATYPE  H5T_STD_I32LE
            DATASPACE  SCALAR
            DATA {
            (0): 2
            }
         }
      }
   }
}
}
```
and
```
HDF5 "test-h5netcdf.nc" {
GROUP "/" {
   GROUP "grp1" {
      DATASET "data1" {
         DATATYPE  H5T_STD_I64LE
         DATASPACE  SIMPLE { ( 2, 2 ) / ( 2, 2 ) }
         DATA {
         (0,0): 1, 2,
         (1,0): 3, 4
         }
         ATTRIBUTE "DIMENSION_LIST" {
            DATATYPE  H5T_VLEN { H5T_REFERENCE { H5T_STD_REF_OBJECT }}
            DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
            DATA {
            (0): (DATASET 1832 /grp1/y ), (DATASET 4480 /grp1/x )
            }
         }
      }
      DATASET "x" {
         DATATYPE  H5T_STD_I64LE
         DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
         DATA {
         (0): 1, 2
         }
         ATTRIBUTE "CLASS" {
            DATATYPE  H5T_STRING {
               STRSIZE 16;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_ASCII;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SCALAR
            DATA {
            (0): "DIMENSION_SCALE"
            }
         }
         ATTRIBUTE "NAME" {
            DATATYPE  H5T_STRING {
               STRSIZE 2;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_ASCII;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SCALAR
            DATA {
            (0): "x"
            }
         }
         ATTRIBUTE "REFERENCE_LIST" {
            DATATYPE  H5T_COMPOUND {
               H5T_REFERENCE { H5T_STD_REF_OBJECT } "dataset";
               H5T_STD_I32LE "dimension";
            }
            DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
            DATA {
            (0): {
                  DATASET 4752 /grp1/data1 ,
                  1
               },
            (1): {
                  DATASET 4752 /grp1/data1 ,
                  1
               }
            }
         }
         ATTRIBUTE "_Netcdf4Dimid" {
            DATATYPE  H5T_STD_I64LE
            DATASPACE  SCALAR
            DATA {
            (0): 1
            }
         }
      }
      DATASET "y" {
         DATATYPE  H5T_STD_I64LE
         DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
         DATA {
         (0): 1, 2
         }
         ATTRIBUTE "CLASS" {
            DATATYPE  H5T_STRING {
               STRSIZE 16;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_ASCII;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SCALAR
            DATA {
            (0): "DIMENSION_SCALE"
            }
         }
         ATTRIBUTE "NAME" {
            DATATYPE  H5T_STRING {
               STRSIZE 2;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_ASCII;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SCALAR
            DATA {
            (0): "y"
            }
         }
         ATTRIBUTE "REFERENCE_LIST" {
            DATATYPE  H5T_COMPOUND {
               H5T_REFERENCE { H5T_STD_REF_OBJECT } "dataset";
               H5T_STD_I32LE "dimension";
            }
            DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
            DATA {
            (0): {
                  DATASET 4752 /grp1/data1 ,
                  0
               },
            (1): {
                  DATASET 4752 /grp1/data1 ,
                  0
               }
            }
         }
         ATTRIBUTE "_Netcdf4Dimid" {
            DATATYPE  H5T_STD_I64LE
            DATASPACE  SCALAR
            DATA {
            (0): 0
            }
         }
      }
   }
   GROUP "grp2" {
      DATASET "data2" {
         DATATYPE  H5T_STD_I64LE
         DATASPACE  SIMPLE { ( 3, 3 ) / ( 3, 3 ) }
         DATA {
         (0,0): 1, 2, 3,
         (1,0): 4, 5, 6,
         (2,0): 7, 8, 9
         }
         ATTRIBUTE "DIMENSION_LIST" {
            DATATYPE  H5T_VLEN { H5T_REFERENCE { H5T_STD_REF_OBJECT }}
            DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
            DATA {
            (0): (DATASET 11328 /grp2/y ), (DATASET 11928 /grp2/x )
            }
         }
      }
      DATASET "x" {
         DATATYPE  H5T_STD_I64LE
         DATASPACE  SIMPLE { ( 3 ) / ( 3 ) }
         DATA {
         (0): 1, 2, 3
         }
         ATTRIBUTE "CLASS" {
            DATATYPE  H5T_STRING {
               STRSIZE 16;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_ASCII;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SCALAR
            DATA {
            (0): "DIMENSION_SCALE"
            }
         }
         ATTRIBUTE "NAME" {
            DATATYPE  H5T_STRING {
               STRSIZE 2;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_ASCII;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SCALAR
            DATA {
            (0): "x"
            }
         }
         ATTRIBUTE "REFERENCE_LIST" {
            DATATYPE  H5T_COMPOUND {
               H5T_REFERENCE { H5T_STD_REF_OBJECT } "dataset";
               H5T_STD_I32LE "dimension";
            }
            DATASPACE  SIMPLE { ( 1 ) / ( 1 ) }
            DATA {
            (0): {
                  DATASET 12200 /grp2/data2 ,
                  1
               }
            }
         }
         ATTRIBUTE "_Netcdf4Dimid" {
            DATATYPE  H5T_STD_I64LE
            DATASPACE  SCALAR
            DATA {
            (0): 1
            }
         }
      }
      DATASET "y" {
         DATATYPE  H5T_STD_I64LE
         DATASPACE  SIMPLE { ( 3 ) / ( 3 ) }
         DATA {
         (0): 1, 2, 3
         }
         ATTRIBUTE "CLASS" {
            DATATYPE  H5T_STRING {
               STRSIZE 16;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_ASCII;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SCALAR
            DATA {
            (0): "DIMENSION_SCALE"
            }
         }
         ATTRIBUTE "NAME" {
            DATATYPE  H5T_STRING {
               STRSIZE 2;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_ASCII;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SCALAR
            DATA {
            (0): "y"
            }
         }
         ATTRIBUTE "REFERENCE_LIST" {
            DATATYPE  H5T_COMPOUND {
               H5T_REFERENCE { H5T_STD_REF_OBJECT } "dataset";
               H5T_STD_I32LE "dimension";
            }
            DATASPACE  SIMPLE { ( 1 ) / ( 1 ) }
            DATA {
            (0): {
                  DATASET 12200 /grp2/data2 ,
                  0
               }
            }
         }
         ATTRIBUTE "_Netcdf4Dimid" {
            DATATYPE  H5T_STD_I64LE
            DATASPACE  SCALAR
            DATA {
            (0): 0
            }
         }
      }
   }
}
}
```

Here's the diff:
```
1c1
< HDF5 "test-netcdf4.nc" {
---
> HDF5 "test.nc" {
3,14d2
<    ATTRIBUTE "_NCProperties" {
<       DATATYPE  H5T_STRING {
<          STRSIZE 34;
<          STRPAD H5T_STR_NULLTERM;
<          CSET H5T_CSET_ASCII;
<          CTYPE H5T_C_S1;
<       }
<       DATASPACE  SCALAR
<       DATA {
<       (0): "version=2,netcdf=4.6.2,hdf5=1.10.4"
<       }
<    }
27,34c15
<             (0): (DATASET 566 /grp1/y ), (DATASET 876 /grp1/x )
<             }
<          }
<          ATTRIBUTE "_Netcdf4Dimid" {
<             DATATYPE  H5T_STD_I32LE
<             DATASPACE  SCALAR
<             DATA {
<             (0): 0
---
>             (0): (DATASET 1832 /grp1/y ), (DATASET 4480 /grp1/x )
73c54
<             DATASPACE  SIMPLE { ( 1 ) / ( 1 ) }
---
>             DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
76c57,61
<                   DATASET 3294 /grp1/data1 ,
---
>                   DATASET 4752 /grp1/data1 ,
>                   1
>                },
>             (1): {
>                   DATASET 4752 /grp1/data1 ,
82c67
<             DATATYPE  H5T_STD_I32LE
---
>             DATATYPE  H5T_STD_I64LE
124c109
<             DATASPACE  SIMPLE { ( 1 ) / ( 1 ) }
---
>             DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
127c112,116
<                   DATASET 3294 /grp1/data1 ,
---
>                   DATASET 4752 /grp1/data1 ,
>                   0
>                },
>             (1): {
>                   DATASET 4752 /grp1/data1 ,
133c122
<             DATATYPE  H5T_STD_I32LE
---
>             DATATYPE  H5T_STD_I64LE
154c143
<             (0): (DATASET 9761 /grp2/y ), (DATASET 10071 /grp2/x )
---
>             (0): (DATASET 11328 /grp2/y ), (DATASET 11928 /grp2/x )
196c185
<                   DATASET 12497 /grp2/data2 ,
---
>                   DATASET 12200 /grp2/data2 ,
202c191
<             DATATYPE  H5T_STD_I32LE
---
>             DATATYPE  H5T_STD_I64LE
205c194
<             (0): 3
---
>             (0): 1
247c236
<                   DATASET 12497 /grp2/data2 ,
---
>                   DATASET 12200 /grp2/data2 ,
253c242
<             DATATYPE  H5T_STD_I32LE
---
>             DATATYPE  H5T_STD_I64LE
256c245
<             (0): 2
---
>             (0): 0
```

**Comment by shoyer** [comment]
To summarize the salient features of the diff:
- netCDF-C produces a file with a `_Netcdf4Dimid` attribute on the `grp1/data1` dataset. This seems a little strange -- this variable definitely isn't a dimension scale.
- h5netcdf produces a file where `grp1/x` and `grp1/y` have `REFERENCE_LIST` attributes of length 2. This also looks suspicious -- which two variables are these dimension scales associated with?

**Comment by shoyer** [comment]
The real issue seems to be reusing dimension IDs.

Fixed by 13dae1afe349a829b681a78f33dc721e2e727689
